shell/platform/darwin/ios/framework/Source/FlutterEngine.mm

```diff
@@ -113,13 +113,13 @@ - (instancetype)initWithName:(NSString*)labelPrefix
                object:nil];
 
   [center addObserver:self
-             selector:@selector(applicationBecameActive:)
-                 name:UIApplicationDidBecomeActiveNotification
+             selector:@selector(applicationWillEnterForeground:)
```

Comment:
I'm not sure how these notifications match access to the GPU.  The problem is the wording "Will Enter Foreground", which means it isn't currently in the foreground, which means it's in the background.  If it's in the background does it have access to the GPU?  Typically no, but the wording on the notification might just be awkward.

Unfortunately it's hard to prove this is correct because even with `setIsGpuDisabled` removed, the crash doesn't manifest 100% of the time.  It's a race condition.

If we could provide instructions that could reproduce the crash 100% of the time with the setIsGpuDisabled removed, we could have more confidence in the change.  I'll try playing around with it.

Reply:
From apples docs on didEnterBackground:

> Use this method to release shared resources, invalidate timers, and store enough app state information to restore your app to its current state in case it's terminated later. **Disable updates to your app’s user interface**, and avoid using some types of shared system resources (such as the user’s contacts database). Don’t use OpenGL ES in the background.

From willEnterForeground:

> In iOS 4.0 and later, UIKit calls this method as part of the transition from the background to the active state. You can use this method to **undo many of the changes you made to your app upon entering the background**.

In the background you should not access the gpu indeed but before `didEnterBackground` is called we’re not in the background. And when `willEnterForeground` we are also no longer in the background.

Also the metal documentation I linked before has this:

> When UIKit calls your app delegate's applicationDidEnterBackground(_:) method, make sure Metal has scheduled all command buffers you've already committed before your app returns control to the system. On each command queue, if the last command buffer you queued isn't already scheduled or complete, call waitUntilScheduled() to force it to be scheduled.

Also the crash is only when using openGL ES
When using metal any commands in the background will simply return an error instead of crashing the app:

> After your app moves into the background, if Metal sees a new command buffer from your app, it returns an error, rather than scheduling the command buffer. You can test for this error by adding a completion handler (addCompletedHandler(_:)). Inside your completion handler, check the status and error properties.

Reply:
> The problem is the wording "Will Enter Foreground", which means it isn't currently in the foreground, which means it's in the background.

This is not correct, the app has more than 2 states.
When willEnterForeground is called it’s no longer in the background but in the inactive state, it might not be visible to the user yet though but that’s independent of the state

Reply:
Finally from `willResignActive` which we’re currently using to stop rendering:

> Use this method to pause ongoing tasks, disable timers, and **throttle down OpenGL ES frame rates**. Games should use this method to pause the game. An app in the inactive state should do minimal work while it waits to transition to either the active or background state.

So in this method the app should not stop rendering just try to use less resources

Hopefully these points help you figure this out 🙂
I’m glad it’s finally being looked at again!